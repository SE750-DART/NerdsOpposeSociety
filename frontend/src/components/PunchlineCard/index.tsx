import React from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faStar } from "@fortawesome/free-solid-svg-icons";
import styles from "./style.module.css";

interface PunchlineCardProps {
  text: string;
  handleOnClick?: () => void;
  status?: "available" | "selected" | "submitted";
  newCard?: boolean;
  selectedNum?: 1 | 2 | 3;
  blurred?: boolean;
}

const PunchlineCard = ({
  text,
  handleOnClick,
  status = "available",
  newCard = false,
  selectedNum,
  blurred = false,
}: PunchlineCardProps) => {
  // Max length of a punchline before it needs to be made smaller to fit
  const textBreakpoint: number = 69;

  let textStyle: string = "";
  if (text.length > textBreakpoint) {
    textStyle = styles.textSmall;
  }

  let cardStyle: string = styles.punchlineAvailable;
  if (status === "selected") {
    cardStyle = styles.punchlineSelected;
  } else if (status === "submitted") {
    cardStyle = styles.punchlineSubmitted;
  }

  return (
    <div
      className={`${styles.punchline} ${cardStyle}`}
      role="button"
      onClick={handleOnClick}
      onKeyDown={handleOnClick}
      tabIndex={0}
    >
      {newCard && (
        <div className={`${styles.indicator} ${styles.newIndicator}`}>
          <FontAwesomeIcon icon={faStar} />
        </div>
      )}
      {selectedNum && (
        <div className={`${styles.indicator} ${styles.numIndicator}`}>
          <h5>{selectedNum}</h5>
        </div>
      )}
      <p
        className={`${textStyle} ${blurred ? styles.textBlurred : undefined}`}
        style={{ transition: "0.3s" }}
      >
        {text}
      </p>
    </div>
  );
};

export default PunchlineCard;
