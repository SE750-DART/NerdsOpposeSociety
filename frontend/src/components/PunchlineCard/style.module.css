.punchline {
  border-radius: 16px;
  box-shadow: 0px 10px 20px rgba(0, 0, 0, 0.1);
  cursor: pointer;
  height: fit-content;
  margin-bottom: 24px;
  max-height: 154px; /* This magic number is the height of the longest punchline */
  min-height: 100px;
  padding: 20px;
  position: relative;
  transition: 0.3s;
}

.punchline:hover {
  box-shadow: 0px 10px 40px rgba(0, 0, 0, 0.25);
  transition: 0.3s;
}

.punchlineAvailable {
  background-color: #eeffff;
  border: 4px solid #eeffff;
}

.punchlineSelected {
  background-color: #ffffff;
  border: 4px solid #84b5de;
  box-shadow: 0px 10px 40px rgba(0, 0, 0, 0.25);
  transition: 0.3s;
}

.punchlineSubmitted {
  background-color: #84b5de;
  box-shadow: inset 0px 10px 20px rgba(0, 0, 0, 0.1);
  cursor: default;
  transition: 0.3s;
}

.punchlineSubmitted:hover {
  box-shadow: inset 0px 10px 20px rgba(0, 0, 0, 0.1);
}

.indicator {
  align-items: center;
  background-color: #84b5de;
  border-radius: 50%;
  display: flex;
  height: 28px;
  justify-content: center;
  position: absolute;
  top: 0;
  width: 28px;
}

.newIndicator {
  left: 0;
  margin: -13px 0 0 -13px;
}

.numIndicator {
  right: 0;
  margin: -13px -13px 0 0;
}

.textSmall {
  font-size: 20px;
}

.textBlurred {
  color: transparent;
  text-shadow: 0 0 10px rgba(0, 0, 0, 0.5);
  transition: 0.6s;
}
