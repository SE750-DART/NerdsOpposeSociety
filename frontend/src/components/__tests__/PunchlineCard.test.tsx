import React from "react";
import renderer from "react-test-renderer";
import PunchlineCard from "../PunchlineCard";

it("Punchline card snapshot testing", () => {
  const shortPunchline: string = "Me time.";
  const longPunchline: string =
    "Looking in the mirror, applying lipstick, and whispering “tonight, you will have sex with Tom Cruise.”";

  const treeBasicPunchline = renderer
    .create(<PunchlineCard text={shortPunchline} />)
    .toJSON();
  expect(treeBasicPunchline).toMatchSnapshot();

  const treeLong = renderer
    .create(<PunchlineCard text={longPunchline} />)
    .toJSON();
  expect(treeLong).toMatchSnapshot();

  const treeNew = renderer
    .create(<PunchlineCard text={shortPunchline} newCard />)
    .toJSON();
  expect(treeNew).toMatchSnapshot();

  const treeNum = renderer
    .create(<PunchlineCard text={shortPunchline} selectedNum={2} />)
    .toJSON();
  expect(treeNum).toMatchSnapshot();
});
