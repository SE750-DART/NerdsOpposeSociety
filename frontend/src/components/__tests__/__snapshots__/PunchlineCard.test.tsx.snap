// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Punchline card snapshot testing 1`] = `
<div
  className="punchline punchlineAvailable"
  role="button"
  tabIndex={0}
>
  <p
    className=" undefined"
    style={
      Object {
        "transition": "0.3s",
      }
    }
  >
    Me time.
  </p>
</div>
`;

exports[`Punchline card snapshot testing 2`] = `
<div
  className="punchline punchlineAvailable"
  role="button"
  tabIndex={0}
>
  <p
    className="textSmall undefined"
    style={
      Object {
        "transition": "0.3s",
      }
    }
  >
    Looking in the mirror, applying lipstick, and whispering “tonight, you will have sex with Tom Cruise.”
  </p>
</div>
`;

exports[`Punchline card snapshot testing 3`] = `
<div
  className="punchline punchlineAvailable"
  role="button"
  tabIndex={0}
>
  <div
    className="indicator newIndicator"
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-star fa-w-18 "
      data-icon="star"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 576 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M259.3 17.8L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
  </div>
  <p
    className=" undefined"
    style={
      Object {
        "transition": "0.3s",
      }
    }
  >
    Me time.
  </p>
</div>
`;

exports[`Punchline card snapshot testing 4`] = `
<div
  className="punchline punchlineAvailable"
  role="button"
  tabIndex={0}
>
  <div
    className="indicator numIndicator"
  >
    <h5>
      2
    </h5>
  </div>
  <p
    className=" undefined"
    style={
      Object {
        "transition": "0.3s",
      }
    }
  >
    Me time.
  </p>
</div>
`;
